### DISCUSSION TITLE
Make useLoaderData and useActionData hooks use type inference by default

### DISCUSSION THREAD

**Discussion by mikeybinns** [remix-run/remix#5668] [Proposals]
For useLoaderData and useActionData, the most common use case is to add a generic of `<typeof loader>` and `<typeof action>` respectively.

I think in Remix v2, these should be made the default for better DX and type safety. We could also add this behind a future config tag, e.g. v2_useDataHooks.

**Comment by kiliman** [comment]
The problem is that there is no way for TypeScript to know which function to infer from.

**Reply by maranomynet** [reply]
A wafer-thin createLoader sugar function wrapper could take a loader function as an argument, infer its return values and return the function and a typed useLoaderData hook.

Something as dumb as this would work:

```ts
const createLoader = <
  L extends (args: LoaderArgs) => Promise<any>
>(loader: L) => ({
  loader,
  useLoaderData: () => useLoaderData<L>()
})
```

Usage:

```ts
const { loader, useLoaderData } = createLoader(asyc (args: LoaderArgs) => {
  return json({ message: 'Hello World' });
})

export const loader;

export default function() {
  const data = useLoaderData();
  return <p>{data.message}</p>;
}
```

**Reply by kiliman** [reply]
The problem there is your `createLoader` HOF is a module side-effect[1], which means your loader code will end up in your client bundle.

The only way to prevent that is to export your loader from `route.server.ts`. It seems like a lot of effort to avoid `useLoaderData<typeof loader>()`

1. https://remix.run/docs/en/1.14.0/guides/constraints#higher-order-functions

**Reply by sergiodxa** [reply]
Another way would be to make the compiler parse the AST and remove it when it found it, there's a proposal which would need to do that, but I think passing the type is simpler.

**Reply by kiliman** [reply]
True. The new `createLoader` proposal would probably make this a non-issue. But until something like that lands, you're kind of limited with what is possible with pure TypeScript (sans bespoke compiler).

**Reply by cliffordfajardo** [reply]
Here is `createLoader` proposal that Kiliman mentioned: https://github.com/remix-run/remix/discussions/5383

**Reply by maranomynet** [reply]
@kiliman thanks for the link.
This is obvious, once you point it out.

**Comment by mikeybinns** [comment]
Ah yeah, you're right 👍

**Comment by sgrishchenko** [comment]
Can we add generics for return type like in `useOutletContext`? I can create PR with such improvement.

**Reply by mikeybinns** [reply]
useLoaderData and useActionData already have generics, and they accept typeof loader and typeof action for inferring the return type. My discussion was about making those values the default (which isn't possible for the reasons stated above)

**Reply by sgrishchenko** [reply]
As far as I can see there is no generics here
https://github.com/remix-run/react-router/blob/main/packages/react-router/lib/hooks.tsx#L784
https://github.com/remix-run/react-router/blob/main/packages/react-router/lib/hooks.tsx#L800
https://github.com/remix-run/react-router/blob/main/packages/react-router/lib/hooks.tsx#L808
Do you mean some branch where such generics are added?

**Reply by mikeybinns** [reply]
I don't know where they are in react router, I have only used Remix, and they are definitely there in Remix. Perhaps someone more knowledgable than myself in that area can point you in the right direction, but for discoverability, I'd probably open a new question, or ask in the Remix discord.
